### PR TITLE
New recursive composition system

### PR DESCRIPTION
**Pull request by imeckler**
This branch contains the new Pickles recursive composition system. It depends on https://github.com/CodaProtocol/coda/pull/4338.

See https://github.com/CodaProtocol/coda/blob/1ce936afd5435f964ee94c96fa15a84efb420c60/src/lib/pickles/pickles.ml#L2480 for an example of how it is used.

### CODE REVIEW

File: src/lib/pickles/dlog_main_inputs.ml [hunk 1]
```diff
@@ -0,0 +1,230 @@
+open Core_kernel
+open Common
+
+(*
```
Comment (imeckler) [inline src/lib/pickles/dlog_main_inputs.ml]: Delete

File: src/lib/pickles/impls.ml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      Typ.transport Field.typ ~there:fp_as_fq ~back:(fun (x : Fq.t) ->
+          Fp.of_bigint (Fq.to_bigint x) )
+    in
+    let open Types.Dlog_based.Statement in
+    let (T (typ, f)) = Spec.packed_typ (module Impl) (T (fp, Fn.id)) spec in
+    let typ = Typ.transport typ ~there:to_data ~back:of_data in
+    Spec.ETyp.T (typ, fun x -> of_data (f x))
+end
+
+(*
```
Comment (imeckler) [inline src/lib/pickles/impls.ml]: delete

File: src/lib/pickles/inductive_rule.ml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    ( 'prev_vars
+    , 'prev_values
+    , 'widths
+    , 'heights
+    , Statement.t
+    , Statement_value.t )
+    t
+end
+
+(*
```
Comment (imeckler) [inline src/lib/pickles/inductive_rule.ml]: delete

File: src/lib/pickles/fixed_map.ml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+struct
+  let order = `Increasing_key
+
+  let keys : M.Key.t list =
+    M.of_alist_exn (List.map Keys.keys ~f:(fun k -> (k, ())))
+    |> M.to_sequence ~order |> Sequence.map ~f:fst |> Sequence.to_list
+
+  type 'a t = 'a M.t [@@deriving sexp]
+
+  include Binable.Of_binable1
```
Comment (psteckler) [inline src/lib/pickles/fixed_map.ml]: this creates a serializable type that's not versioned

File: src/lib/pickles/intf.ml [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+  end
+end
+
+module Evals = struct
+  module type S = sig
+    type n
+
+    val n : n Vector.nat
+
+    include Binable.S1 with type 'a t = ('a, n) Vector.t
```
Comment (psteckler) [inline src/lib/pickles/intf.ml]: probably this should be expressed with `%%versioned:` (and the type with this signature should be versioned)

File: src/lib/pickles/pairing_main_inputs.ml [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+      let group_size_in_bits = 382
+    end
+
+    module F = struct
+      include struct
+        open Impl.Field
+
+        type nonrec t = t
+
+        let ( * ), ( + ), ( - ), inv_exn, square, scale, if_, typ, constant =
```
Comment (psteckler) [inline src/lib/pickles/pairing_main_inputs.ml]: I should probably extend `define_locally` to handle this case.

File: src/lib/pickles/pickles.mli [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  val to_field_elements : t -> field array
+end
+
+module type Statement_var_intf =
+  Statement_intf with type field := Impls.Pairing_based.Field.t
+
+module type Statement_value_intf = sig
+  include Statement_intf with type field := Impls.Pairing_based.field
+
+  include Binable.S with type t := t
```
Comment (psteckler) [inline src/lib/pickles/pickles.mli]: should also use `%%versioned:`

File: src/lib/pickles/types.ml [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+      module Marlin = Dlog_based.Proof_state.Deferred_values.Marlin
+
+      type ('challenge, 'fq, 'bulletproof_challenges) t =
+        { marlin: ('challenge, 'fq) Marlin.t
+        ; combined_inner_product: 'fq
+        ; xi: 'challenge (* 128 bits *)
+        ; r: 'challenge (* 128 bits *)
+        ; bulletproof_challenges: 'bulletproof_challenges
+        ; b: 'fq }
+      [@@deriving bin_io]
```
Comment (psteckler) [inline src/lib/pickles/types.ml]: because it has `bin_io`, should be a versioned type

File: src/lib/pickles/pairing_main_inputs.ml (hunk 6 above)
Comment (psteckler) [reply src/lib/pickles/pairing_main_inputs.ml]: When #4588 is merged, you'll be able to write:
```ocaml
[%%define_from_scope (( * ), ( + ), ( - ), inv_exn, square, scale, if_, typ, constant)]
```

**Pull request closed without merging**